lib/Dialect/FIRRTL/FIRRTLInstanceImplementation.cpp

```diff
@@ -0,0 +1,137 @@
+  if (portAnnotations.size() != numExpected)
+    return emitNote(
+        instanceOp->emitOpError("the number of result annotations should be "
+                                "equal to the number of results"));
+
+  // Check that the port names match the referenced module.
+  if (portNames != referencedModule.getPortNamesAttr()) {
+    // We know there is an error, try to figure out whats wrong.
+    auto moduleNames = referencedModule.getPortNamesAttr();
+    // First compare the sizes:
+    if (portNames.size() != moduleNames.size()) {
+      return emitNote(instanceOp->emitOpError()
+                      << "has a wrong number of directions; expected "
+                      << moduleNames.size() << " but got " << portNames.size());
+    }
+    // Next check the values:
+    for (size_t i = 0; i != numResults; ++i) {
+      if (portNames[i] != moduleNames[i]) {
+        return emitNote(instanceOp->emitOpError()
+                        << "name for port " << i << " must be "
+                        << moduleNames[i] << ", but got " << portNames[i]);
+      }
+    }
+    llvm_unreachable("should have found something wrong");
+  }
+
+  // Check that the types match.
+  for (size_t i = 0; i != numResults; i++) {
+    auto resultType = instanceOp->getResult(i).getType();
+    auto expectedType = referencedModule.getPortType(i);
+    if (resultType != expectedType) {
+      return emitNote(instanceOp->emitOpError()
+                      << "result type for " << portNames[i] << " must be "
+                      << expectedType << ", but got " << resultType);
+    }
+  }
+
+  // Check that the port directions are consistent with the referenced module's.
+  if (portDirections != referencedModule.getPortDirectionsAttr()) {
+    // We know there is an error, try to figure out whats wrong.
+    auto moduleDirectionAttr = referencedModule.getPortDirectionsAttr();
+    // First compare the sizes:
+    auto expectedWidth = moduleDirectionAttr.getValue().getBitWidth();
+    auto actualWidth = portDirections.getValue().getBitWidth();
+    if (expectedWidth != actualWidth) {
+      return emitNote(instanceOp->emitOpError()
+                      << "has a wrong number of directions; expected "
+                      << expectedWidth << " but got " << actualWidth);
+    }
+    // Next check the values.
+    auto instanceDirs = direction::unpackAttribute(portDirections);
+    auto moduleDirs = direction::unpackAttribute(moduleDirectionAttr);
+    for (size_t i = 0; i != numResults; ++i) {
+      if (instanceDirs[i] != moduleDirs[i]) {
+        return emitNote(instanceOp->emitOpError()
+                        << "direction for " << portNames[i] << " must be \""
+                        << direction::toString(moduleDirs[i])
+                        << "\", but got \""
+                        << direction::toString(instanceDirs[i]) << "\"");
+      }
+    }
+    llvm_unreachable("should have found something wrong");
```

Comment:
nit: we shouldn't use llvm_unreachable https://github.com/llvm/circt/issues/3836